File: backend/src/main/java/com/interview/service/AdminService.java
package com.interview.service;

import com.interview.dto.UserProfileDTO;
import com.interview.exception.BadRequestException;
import com.interview.exception.ResourceNotFoundException;
import com.interview.model.Interview;
import com.interview.model.User;
import com.interview.repository.InterviewRepository;
import com.interview.repository.UserRepository;
import com.interview.util.ModelMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.Map;

/**
 * Service for admin operations
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AdminService {

    private final UserRepository userRepository;
    private final InterviewRepository interviewRepository;
    private final ModelMapper modelMapper;

    public Page<UserProfileDTO> getAllUsers(Pageable pageable) {
        log.info("Getting all users");
        return userRepository.findAll(pageable)
                .map(modelMapper::toUserProfileDTO);
    }

    public Page<UserProfileDTO> searchUsers(String email, Pageable pageable) {
        log.info("Searching users by email: {}", email);
        return userRepository.findByEmailContainingIgnoreCase(email, pageable)
                .map(modelMapper::toUserProfileDTO);
    }

    public UserProfileDTO getUserById(String userId) {
        log.info("Getting user by ID: {}", userId);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));
        return modelMapper.toUserProfileDTO(user);
    }

    @Transactional
    public void suspendUser(String userId) {
        log.info("Suspending user: {}", userId);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));

        if (user.getRoles().contains(User.Role.ADMIN)) {
            throw new BadRequestException("Cannot suspend admin users");
        }

        user.setStatus(User.UserStatus.SUSPENDED);
        userRepository.save(user);
        log.info("User suspended successfully");
    }

    @Transactional
    public void activateUser(String userId) {
        log.info("Activating user: {}", userId);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));

        user.setStatus(User.UserStatus.ACTIVE);
        userRepository.save(user);
        log.info("User activated successfully");
    }

    @Transactional
    public void deleteUser(String userId) {
        log.info("Deleting user: {}", userId);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));

        if (user.getRoles().contains(User.Role.ADMIN)) {
            throw new BadRequestException("Cannot delete admin users");
        }

        user.setStatus(User.UserStatus.DELETED);
        userRepository.save(user);
        log.info("User deleted successfully");
    }

    public Map<String, Object> getPlatformAnalytics() {
        log.info("Getting platform analytics");

        Map<String, Object> analytics = new HashMap<>();

        // User statistics
        long totalUsers = userRepository.count();
        long activeUsers = userRepository.countByStatus(User.UserStatus.ACTIVE);
        long suspendedUsers = userRepository.countByStatus(User.UserStatus.SUSPENDED);

        analytics.put("totalUsers", totalUsers);
        analytics.put("activeUsers", activeUsers);
        analytics.put("suspendedUsers", suspendedUsers);

        // Interview statistics
        long totalInterviews = interviewRepository.count();
        long completedInterviews = interviewRepository.countAll()
                .stream()
                .filter(interview -> interview.getStatus() == Interview.InterviewStatus.COMPLETED)
                .count();
        long inProgressInterviews = interviewRepository.countAll()
                .stream()
                .filter(interview -> interview.getStatus() == Interview.InterviewStatus.IN_PROGRESS)
                .count();

        analytics.put("totalInterviews", totalInterviews);
        analytics.put("completedInterviews", completedInterviews);
        analytics.put("inProgressInterviews", inProgressInterviews);

        // Calculate average score across all completed interviews
        double avgScore = interviewRepository.findAll().stream()
                .filter(interview -> interview.getStatus() == Interview.InterviewStatus.COMPLETED)
                .filter(interview -> interview.getOverallScore() != null)
                .mapToDouble(Interview::getOverallScore)
                .average()
                .orElse(0.0);

        analytics.put("platformAverageScore", avgScore);

        // Domain distribution
        Map<String, Long> domainDistribution = new HashMap<>();
        for (Interview.InterviewDomain domain : Interview.InterviewDomain.values()) {
            long count = interviewRepository.findAll().stream()
                    .filter(interview -> interview.getDomain() == domain)
                    .count();
            domainDistribution.put(domain.name(), count);
        }
        analytics.put("domainDistribution", domainDistribution);

        return analytics;
    }

    public Map<String, Object> getAIUsageStatistics() {
        log.info("Getting AI usage statistics");

        Map<String, Object> stats = new HashMap<>();

        // Total AI API calls (approximate: total questions + total answers)
        long totalInterviews = interviewRepository.count();
        long estimatedApiCalls = totalInterviews * 10; // Assuming avg 10 questions per interview

        stats.put("totalInterviews", totalInterviews);
        stats.put("estimatedApiCalls", estimatedApiCalls);

        // Average questions per interview
        double avgQuestionsPerInterview = interviewRepository.findAll().stream()
                .filter(interview -> interview.getTotalQuestions() != null)
                .mapToInt(Interview::getTotalQuestions)
                .average()
                .orElse(0.0);

        stats.put("avgQuestionsPerInterview", avgQuestionsPerInterview);

        // Most used difficulty level
        Map<String, Long> difficultyDistribution = new HashMap<>();
        for (Interview.Difficulty difficulty : Interview.Difficulty.values()) {
            long count = interviewRepository.findAll().stream()
                    .filter(interview -> interview.getDifficulty() == difficulty)
                    .count();
            difficultyDistribution.put(difficulty.name(), count);
        }
        stats.put("difficultyDistribution", difficultyDistribution);

        return stats;
    }
}

File: backend/src/services/admin.service.js
const User = require('../models/user.model');
const Interview = require('../models/interview.model');
const { NotFoundError, BadRequestError } = require('../utils/errors');
const modelMapper = require('../utils/model-mapper');

const getAllUsers = async (page = 0, size = 10) => {
  console.log('Getting all users paginated');
  
  const totalElements = await User.countDocuments();
  const users = await User.find()
    .skip(page * size)
    .limit(size);
    
  const totalPages = Math.ceil(totalElements / size);
  const content = users.map(modelMapper.toUserProfileDTO);
  
  return {
    content,
    pageable: {
      pageNumber: page,
      pageSize: size
    },
    totalElements,
    totalPages,
    last: page >= totalPages - 1,
    first: page === 0,
    size,
    number: page,
    numberOfElements: content.length,
    empty: content.length === 0
  };
};

const searchUsers = async (email, page = 0, size = 10) => {
  console.log(`Searching users by email: ${email}`);
  
  const query = { email: { $regex: email, $options: 'i' } };
  const totalElements = await User.countDocuments(query);
  const users = await User.find(query)
    .skip(page * size)
    .limit(size);
    
  const totalPages = Math.ceil(totalElements / size);
  const content = users.map(modelMapper.toUserProfileDTO);
  
  return {
    content,
    pageable: {
      pageNumber: page,
      pageSize: size
    },
    totalElements,
    totalPages,
    last: page >= totalPages - 1,
    first: page === 0,
    size,
    number: page,
    numberOfElements: content.length,
    empty: content.length === 0
  };
};

const getUserById = async (userId) => {
  console.log(`Getting user by ID: ${userId}`);
  const user = await User.findById(userId);
  if (!user) {
    throw new NotFoundError('User not found');
  }
  return modelMapper.toUserProfileDTO(user);
};

const suspendUser = async (userId) => {
  console.log(`Suspending user: ${userId}`);
  const user = await User.findById(userId);
  if (!user) {
    throw new NotFoundError('User not found');
  }

  if (user.roles.includes('ADMIN')) {
    throw new BadRequestError('Cannot suspend admin users');
  }

  user.status = 'SUSPENDED';
  await user.save();
  console.log('User suspended successfully');
};

const activateUser = async (userId) => {
  console.log(`Activating user: ${userId}`);
  const user = await User.findById(userId);
  if (!user) {
    throw new NotFoundError('User not found');
  }

  user.status = 'ACTIVE';
  await user.save();
  console.log('User activated successfully');
};

const deleteUser = async (userId) => {
  console.log(`Deleting user: ${userId}`);
  const user = await User.findById(userId);
  if (!user) {
    throw new NotFoundError('User not found');
  }

  if (user.roles.includes('ADMIN')) {
    throw new BadRequestError('Cannot delete admin users');
  }

  user.status = 'DELETED';
  await user.save();
  console.log('User deleted successfully');
};

const getPlatformAnalytics = async () => {
  console.log('Getting platform analytics');

  const totalUsers = await User.countDocuments();
  const activeUsers = await User.countDocuments({ status: 'ACTIVE' });
  const suspendedUsers = await User.countDocuments({ status: 'SUSPENDED' });

  const interviews = await Interview.find();
  const totalInterviews = interviews.length;
  
  const completedInterviews = interviews.filter(i => i.status === 'COMPLETED').length;
  const inProgressInterviews = interviews.filter(i => i.status === 'IN_PROGRESS').length;

  const completedWithScore = interviews.filter(i => i.status === 'COMPLETED' && i.overallScore !== undefined && i.overallScore !== null);
  const avgScore = completedWithScore.length > 0
    ? completedWithScore.reduce((sum, i) => sum + i.overallScore, 0) / completedWithScore.length
    : 0.0;

  // Domain distribution
  const domains = ['JAVA', 'SPRING_BOOT', 'MERN', 'DSA', 'HR', 'SQL', 'JAVASCRIPT', 'CODEFORCES'];
  const domainDistribution = {};
  domains.forEach(domain => {
    domainDistribution[domain] = interviews.filter(i => i.domain === domain).length;
  });

  return {
    totalUsers,
    activeUsers,
    suspendedUsers,
    totalInterviews,
    completedInterviews,
    inProgressInterviews,
    platformAverageScore: avgScore,
    domainDistribution
  };
};

const getAIUsageStatistics = async () => {
  console.log('Getting AI usage statistics');

  const totalInterviews = await Interview.countDocuments();
  const estimatedApiCalls = totalInterviews * 10; // Assuming avg 10 API calls per interview

  const interviews = await Interview.find({ totalQuestions: { $exists: true } });
  const avgQuestionsPerInterview = interviews.length > 0
    ? interviews.reduce((sum, i) => sum + i.totalQuestions, 0) / interviews.length
    : 0.0;

  // Difficulty distribution
  const difficulties = ['EASY', 'MEDIUM', 'HARD'];
  const difficultyDistribution = {};
  difficulties.forEach(diff => {
    difficultyDistribution[diff] = interviews.filter(i => i.difficulty === diff).length;
  });

  return {
    totalInterviews,
    estimatedApiCalls,
    avgQuestionsPerInterview,
    difficultyDistribution
  };
};

module.exports = {
  getAllUsers,
  searchUsers,
  getUserById,
  suspendUser,
  activateUser,
  deleteUser,
  getPlatformAnalytics,
  getAIUsageStatistics
};
